dig.go

```diff
@@ -57,10 +57,58 @@ type optionFunc func(*Container)
 
 func (f optionFunc) applyOption(c *Container) { f(c) }
 
-// A ProvideOption modifies the default behavior of Provide. It's included for
-// future functionality; currently, there are no concrete implementations.
+type provideOptions struct {
+	Name string
+}
+
+func (o *provideOptions) Validate() error {
+	// Names must be representable inside a backquoted string. The only
+	// limitation for raw string literals as per
+	// https://golang.org/ref/spec#raw_string_lit is that they cannot contain
+	// backquotes.
+	if strings.ContainsRune(o.Name, '`') {
+		return fmt.Errorf("invalid dig.Name(%q): names cannot contain backquotes", o.Name)
+	}
+	return nil
+}
+
+// A ProvideOption modifies the default behavior of Provide.
 type ProvideOption interface {
-	unimplemented()
+	applyProvideOption(*provideOptions)
+}
+
+type provideOptionFunc func(*provideOptions)
+
+func (f provideOptionFunc) applyProvideOption(opts *provideOptions) { f(opts) }
+
+// Name is a ProvideOption that specifies that all values produced by a
+// constructor should have the given name. See also the package documentation
+// about Named Values.
+//
+// The name also applies to the fields of a dig.Out struct returned by the
+// constructor if those fields did not have their own name:"..." tag.
```

Comment:
I thought we were going to make this an error? I'd prefer to start with *only* supporting fields that don't have an existing `name` tag.

Reply:
Discussed offline: We will disallow `dig.Out`s for now.

Applying to fields of a `dig.Out` brings up questions we don't have good
answers to yet. For example, if we don't stomp on existing names, it's
impossible for someone to provide two copies of `somemodule.Result` with
different names.

Reply:
Fair enough. Changing this decision isn't backward-compatible, so we may want to call this option `ShallowName` or something similar.

Reply:
Going from error-to-no-error is backwards compatible here. If the following
stops exploding between minor versions, that's completely okay.

```go
c.Provide(NewThing, dig.Name("foo"))
```

The other aspect of the future change is how `name:".."` is interpreted on
embedded result fields. The complexity of behavior there is what lead to
deferring the decision for now by disallowing `dig.Name` on dig.Out structs.

Given,

```go
type Result struct {
  dig.Out
  
  A   string        `name:"a"`
  Foo AnotherResult `name:"foo"`
}

type AnotherResult struct {
    dig.Out
    
    B string `name:"b"`
}
```

The `name:"foo"` on `Result.Foo` has no effect currently because `name:".."`
tags on embedded result structs are ignored.

If we implement some sort of merging behavior here, that could technically
count as breaking change because previously, `Provide(New)` would result in
two values being provided: `string[name="a"]` and `string[name="b"]`. With
merging, the latter may become something like `string[name="foo/b"]`.

However, I think that this is not *actually* breaking because `name:".."` on
embedded result structs is currently unsupported/invalid usage.

Reply:
You're absolutely right. I didn't read carefully enough, and thought that we were *ignoring* result structs - as long as we're erroring, this sounds great.